pkgpanda/util.py

```diff
@@ -279,3 +282,74 @@ def __exit__(self, exc_type, exc_value, traceback):
 def resources_test_dir(path):
     assert not path.startswith('/')
     return "pkgpanda/test_resources/{}".format(path)
+
+
+class MessageLogger:
+    """Abstraction over TeamCity Build Messages
+
+    When pkgpanda is ran in a TeamCity environment additional meta-messages will be output to stdout
+    such that TeamCity can provide improved status reporting, log line highlighting, and failure
+    reporting. When pkgpanda is ran in an environment other than TeamCity all meta-messages will
+    silently be omitted.
+
+    TeamCity docs: https://confluence.jetbrains.com/display/TCD10/Build+Script+Interaction+with+TeamCity
+    """
+    def __init__(self):
+        self.loggers = []
+        if teamcity.is_running_under_teamcity():
+            self.loggers.append(TeamcityServiceMessages())
+        else:
+            self.loggers.append(PrintLogger())
+
+    def _custom_message(self, text, status, error_details='', flow_id=None):
+        for log in self.loggers:
+            log.customMessage(text, status, errorDetails=error_details, flowId=flow_id)
+
+    @contextmanager
+    def _block(self, log, name, flow_id):
+        log.blockOpened(name, flowId=flow_id)
+        log.progressMessage(name)
+        yield
+        log.blockClosed(name, flowId=flow_id)
+
+    @contextmanager
+    def scope(self, name, flow_id=None):
```

Comment:
The idea I had here is that in the code it would read as "With this logging scope run..."  The idea being that all stuff that happened within the block would be happening in that log scope.